Refactor initializers to take VaadinContext instead of ServletContext

jhult opened the pull request:
This PR primarily addresses #5735 but I believe it also solves #2775.

This provides a good step towards allowing external libraries to reduce the dependency on `servlet-api` (see #4613). 

- Added 2 files:
	- `VaadinInitializerException` - a `RuntimeException`
    - `VaadinContextInitializer` - an interface which is implemented by `AnnotationValidator`, `ConnectEndpointsValidator`, `ErrorNavigationTargetInitializer`, `RouteRegisteryInitializer`, `WebComponentConfigurationRegistryInitializer`, `WebComponentExporterAwareValidator`

	For the above 2 files that I added, I picked names that I thought made sense but they can change as need be.

- Refactored initializers to wrap `ServletContext` with `VaadinServletContext`.

	I marked the methods accepting `ServletContext` as deprecated. I think we should do this and encourage people to use the new methods by wrapping `ServletContext` with `VaadinServletContext`.

Questions
1. Should `VaadinInitializerException` extend a `RuntimeException` or `Exception`?

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/flow/8469)
<!-- Reviewable:end -->

pleku commented:
Thanks for the contribution. Sorry for the very slow response.

I've requested review from @denis-anisimov as he has been involved with some related parts quite recently and I'd like to hear his opinion. He won't be back until beginning of September though. I've also restarted the build since it had failed but the results have been cleaned already.

Couple of comments:
- please keep `@since` tags empty
- please rebase the PR on top of master

jhult commented:
Thanks, @pleku.

I see lots of errors in the TeamCity build but they looked related to GWT.

I just rebased (to the `master` branch) and also emptied `@since`.

caalador commented:
```
[03:59:04]	[Step 5/11] [ERROR] javax.servlet.ServletException: java.lang.IllegalArgumentException: argument type mismatch
[03:59:04]	[Step 5/11] [ERROR] 	at com.vaadin.flow.server.startup.ClassLoaderAwareServletContainerInitializer.onStartup(ClassLoaderAwareServletContainerInitializer.java:99)
```

https://teamcity.vaadin.com/viewLog.html?buildId=65788&buildTypeId=Flow_FlowTests&tab=buildLog&_focus=19962

denis-anisimov reviewed (changes requested):
> Reviewed 2 of 12 files at r1, 16 of 16 files at r2.
> *[Reviewable](https://reviewable.io/reviews/vaadin/flow/8469)* status: 17 unresolved discussions, 0 of 1 LGTMs obtained (waiting on @denis-anisimov and @jhult)

___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/AnnotationValidator.java, line 40 at r2](https://reviewable.io/reviews/vaadin/flow/8469#-MG7D3WT4QGPIW7k-Kn5:-MG7D3WU8mWvNKD3QtkQ:b-pufor0) ([raw file](https://github.com/vaadin/flow/blob/c64637e6506472610709bcc91da15cb39e524df5/flow-server/src/main/java/com/vaadin/flow/server/startup/AnnotationValidator.java#L40)):*
> ```
>  ClassLoaderAwareServletContainerInitializer
> ```

Look like this class may extend `VaadinContextInitializer` directly and all deprecation can be done there instead modifying each class which extends `ClassLoaderAwareServletContainerInitializer`.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/ClassLoaderAwareServletContainerInitializer.java, line 90 at r2](https://reviewable.io/reviews/vaadin/flow/8469#-MG7DWyDDtEQxdM_xkJf:-MG7DWyDDtEQxdM_xkJg:b-toqrp9) ([raw file](https://github.com/vaadin/flow/blob/c64637e6506472610709bcc91da15cb39e524df5/flow-server/src/main/java/com/vaadin/flow/server/startup/ClassLoaderAwareServletContainerInitializer.java#L90)):*
> ```
> Stream.of(initializer.getDeclaredMethod(processMethodName, Set.class, ServletContext.class))
> ```

This construction is ugly.
There is no need to use `Stream` if the method returns only one method.
In fact `Stream` has been used to avoid an exception which will be thrown if there is no such methods.
Generally : it's better to avoid an exception if it's possible.
So I would suggest to revert this code back and avoid the exception.
Just check all the methods and choose the required one.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/ConnectEndpointsValidator.java, line 41 at r2](https://reviewable.io/reviews/vaadin/flow/8469#-MG7E8ek5OpOmfOZgxPc:-MG7E8ek5OpOmfOZgxPd:b-8ublg2) ([raw file](https://github.com/vaadin/flow/blob/c64637e6506472610709bcc91da15cb39e524df5/flow-server/src/main/java/com/vaadin/flow/server/startup/ConnectEndpointsValidator.java#L41)):*
> ```
> ClassLoaderAwareServletContainerInitializer
> ```

Here and everywhere else (as above): `ClassLoaderAwareServletContainerInitializer` should extend `VaadinContextInitializer` directly.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java, line 196 at r2](https://reviewable.io/reviews/vaadin/flow/8469#-MG7EWl54ytUHk-qZlII:-MG7EWl60bK2w8d0mt6o:b-mfksdb) ([raw file](https://github.com/vaadin/flow/blob/c64637e6506472610709bcc91da15cb39e524df5/flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java#L196)):*
> ```
> DEV_MODE_HANDLER_ALREADY_STARTED_ATTRIBUTE
> ```


There is no any need in this constant in case of using `DevModeInitializer.class`. It can be used inline.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java, line 600 at r2](https://reviewable.io/reviews/vaadin/flow/8469#-MG7ErX-Aj3-Jsbk_dsW:-MG7ErX-Aj3-Jsbk_dsX:b-n63kuz) ([raw file](https://github.com/vaadin/flow/blob/c64637e6506472610709bcc91da15cb39e524df5/flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java#L600)):*
> ```
> VaadinInitializerException
> ```

The exception should be checked and declared explicitly in the `throws` of the method .
It should be just consistent with `ServletException`.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinAppShellInitializer.java, line 72 at r2](https://reviewable.io/reviews/vaadin/flow/8469#-MG7FDf56CElShXeNocs:-MG7FDf56CElShXeNoct:bmwu7sv) ([raw file](https://github.com/vaadin/flow/blob/c64637e6506472610709bcc91da15cb39e524df5/flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinAppShellInitializer.java#L72)):*
> ```
> VaadinAppShellInitializer
> ```

any reason why this class doesn't implement `VaadinContextInitializer` ?
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinContextInitializer.java, line 43 at r2](https://reviewable.io/reviews/vaadin/flow/8469#-MG7FRIQ0iXdqCmV6rIz:-MG7FRIQ0iXdqCmV6rJ-:b-oy1hcf) ([raw file](https://github.com/vaadin/flow/blob/c64637e6506472610709bcc91da15cb39e524df5/flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinContextInitializer.java#L43)):*
> ```
> process
> ```

Minor :
this method may have another name by the way so that it's easier to find it via reflection.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinInitializerException.java, line 23 at r2](https://reviewable.io/reviews/vaadin/flow/8469#-MG7FbiIC0nn2QuA_8Ra:-MG7FbiIC0nn2QuA_8Rb:b-96s9si) ([raw file](https://github.com/vaadin/flow/blob/c64637e6506472610709bcc91da15cb39e524df5/flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinInitializerException.java#L23)):*
> ```
> RuntimeException
> ```

As mentioned above: this exception should be checked.


<!-- Sent from Reviewable.io -->

jhult commented:
I updated `VaadinContextInitializer` to extend `ClassLoaderAwareServletContainerInitializer` and have a default method which is deprecated.

Then I implemented `VaadinContextInitializer` (instead of `ClassLoaderAwareServletContainerInitializer`) on all but these two classes:

- `DevModeInitializer`
- `VaadinAppShellInitializer`

Both of those eventually need `DeploymentConfiguration` which can't be easily created from the existing method parameters.

I also made `VaadinInitializerException` a checked `Exception`.

Finally, I tried to simplify the method call in `ClassLoaderAwareServletContainerInitializer`.

denis-anisimov commented on flow-server/src/main/java/com/vaadin/flow/server/startup/ClassLoaderAwareServletContainerInitializer.java:
```diff
@@ -78,49 +75,40 @@ default void onStartup(Set<Class<?>> set, ServletContext ctx)
-                    .findFirst().get();
-            operation.invoke(initializer.newInstance(),
-                    new Object[] { set, ctx });
+            Class<? extends ClassLoaderAwareServletContainerInitializer> initializer =
```
Please just revert everything back.
This code makes invalid assumption that `ClassLoaderAwareServletContainerInitializer` `initializer` is loaded by the same classloader .
This is exactly why this code exists.

jhult replied:
This has been reverted.

denis-anisimov commented on flow-server/src/main/java/com/vaadin/flow/server/startup/ClassLoaderAwareServletContainerInitializer.java:
```diff
@@ -107,20 +107,20 @@ default void onStartup(Set<Class<?>> set, ServletContext ctx)
      */
     void process(Set<Class<?>> set, ServletContext ctx) throws ServletException;
-}
+}
```
new line

jhult replied:
New line added at the end of the file.

denis-anisimov commented:
There are a number of GWT unit test failures....
I'm not sure whether they are related to this PR or not: technically it doesn't changes anything on the client side.

jhult commented:
I also see the GWT unit test failures but am unsure how/if they relate to this.

caalador commented:
```
[16:31:34]W:	 [Step 5/11] [ERROR] java.lang.RuntimeException: javax.servlet.ServletException: java.lang.IllegalArgumentException: argument type mismatch
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.annotations.ServletContainerInitializersStarter.doStart(ServletContainerInitializersStarter.java:70)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:274)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1349)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1342)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:741)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:505)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:132)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:114)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:61)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.server.handler.RequestLogHandler.doStart(RequestLogHandler.java:140)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:132)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.server.Server.start(Server.java:387)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:114)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:61)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.server.Server.doStart(Server.java:354)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at com.google.gwt.dev.shell.jetty.JettyLauncher.start(JettyLauncher.java:760)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at com.google.gwt.dev.DevMode.doStartUpServer(DevMode.java:636)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at com.google.gwt.dev.DevModeBase.startUp(DevModeBase.java:923)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at com.google.gwt.junit.JUnitShell.getUnitTestShell(JUnitShell.java:708)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at com.google.gwt.junit.JUnitShell.runTest(JUnitShell.java:682)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at com.google.gwt.junit.client.GWTTestCase.runTest(GWTTestCase.java:421)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at junit.framework.TestCase.runBare(TestCase.java:141)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at junit.framework.TestResult$1.protect(TestResult.java:122)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at junit.framework.TestResult.runProtected(TestResult.java:142)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at junit.framework.TestResult.run(TestResult.java:125)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at junit.framework.TestCase.run(TestCase.java:129)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at com.google.gwt.junit.client.GWTTestCase.run(GWTTestCase.java:247)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at junit.framework.TestSuite.runTest(TestSuite.java:252)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at junit.framework.TestSuite.run(TestSuite.java:247)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at junit.framework.TestSuite.runTest(TestSuite.java:252)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at junit.framework.TestSuite.run(TestSuite.java:247)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.codehaus.mojo.gwt.test.MavenTestRunner.doRun(MavenTestRunner.java:105)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at junit.textui.TestRunner.start(TestRunner.java:183)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.codehaus.mojo.gwt.test.MavenTestRunner.main(MavenTestRunner.java:63)
[16:31:34]W:	 [Step 5/11] [ERROR] Caused by: javax.servlet.ServletException: java.lang.IllegalArgumentException: argument type mismatch
[16:31:34]W:	 [Step 5/11] [ERROR] 	at com.vaadin.flow.server.startup.ClassLoaderAwareServletContainerInitializer.onStartup(ClassLoaderAwareServletContainerInitializer.java:99)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.plus.annotation.ContainerInitializer.callStartup(ContainerInitializer.java:140)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at org.eclipse.jetty.annotations.ServletContainerInitializersStarter.doStart(ServletContainerInitializersStarter.java:65)
[16:31:34]W:	 [Step 5/11] [ERROR] 	... 37 more
[16:31:34]W:	 [Step 5/11] [ERROR] Caused by: java.lang.IllegalArgumentException: argument type mismatch
[16:31:34]W:	 [Step 5/11] [ERROR] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at java.lang.reflect.Method.invoke(Method.java:498)
[16:31:34]W:	 [Step 5/11] [ERROR] 	at com.vaadin.flow.server.startup.ClassLoaderAwareServletContainerInitializer.onStartup(ClassLoaderAwareServletContainerInitializer.java:94)
[16:31:34]W:	 [Step 5/11] [ERROR] 	... 39 more
```

jhult commented:
Thanks @caalador for finding that exception.

@denis-anisimov, that exception is why I originally altered `ClassLoaderAwareServletContainerInitializer` to use this method (which I reverted per your request): `initializer.getDeclaredMethod(processMethodName, Set.class, ServletContext.class)`.

Let me know if you need any additional changes here.

jhult commented:
Is anything else needed on this before it gets merged?

denis-anisimov commented:
I think yes: verification has to pass to be able to merge this.

jhult commented:
@caalador and @denis-anisimov,

I updated the code again so that it is passing the test which was failing (`GwtApplicationConnectionTest`). This is similar to my original fix, but I removed the usage of `Stream`. I'm open to other suggestions on how to fix this if my solution is disliked.

denis-anisimov reviewed (commented):
Technically the current state of the latest changes can be kept as is except `NoSuchMethodException`  should be handled separately and it should throw some exception which informs about implementation error: the method has to be found and something is fundamentally broken if it's thrown.
Previously the absence of the method would have been reported via exception in `get` method.

What I absolutely don't understand: why these changes are needed ?

denis-anisimov commented:
> @caalador and @denis-anisimov,
> 
> I updated the code again so that it is passing the test which was failing (`GwtApplicationConnectionTest`). This is similar to my original fix, but I removed the usage of `Stream`. I'm open to other suggestions on how to fix this if my solution is disliked.

I don't understand two things:
- Why `GwtApplicationConnectionTest` was failing ?
- Why the changes in `ClassLoaderAwareServletContainerInitializer` are required ?

Was  `GwtApplicationConnectionTest` failing because of `ClassLoaderAwareServletContainerInitializer` ? 
In which way?
If not then why it's necessary to touch `ClassLoaderAwareServletContainerInitializer` ?

pleku commented:
This has been hanging for a long time and has merge conflicts. I think removing the dependency to servlet API has merit but it is not a priority at the moment. So it is up for @jhult to handle the conflicts and if the build gets green, then this can be accepted.

knoobie commented:
This addition would be nice because of the upcoming javax go jakarta namespace change. This would allow to run Vaadin with both namespaces if flow can provide to implementations that can be changed depending on the servlet container in use.

jhult commented:
I think I finally sorted out the issue with the test failing.

For the moment, I rebased on top of some work @denis-anisimov is doing (in https://github.com/vaadin/flow/pull/9642) since this code interacts with that in a few places. When that code gets merged to `master`, I can rebase on `master`.

jhult commented on flow-server/src/main/java/com/vaadin/flow/server/startup/ClassLoaderAwareServletContainerInitializer.java:
```diff
@@ -48,7 +48,7 @@
             throws ServletException {
-        // see DeferredServletContextIntializers
+        // see DeferredServletContextInitializers
         DeferredServletContextInitializers.Initializer deferredInitializer = ctx -> {
```
This existed prior to me. All I did was fix a typo.

jhult commented on flow-server/src/main/java/com/vaadin/flow/server/startup/LookupInitializer.java:
```diff
@@ -200,22 +201,21 @@ public InputStream getClientResourceAsStream(String path)
 
         DeferredServletContextInitializers initializers;
-        synchronized (servletContext) {
+        synchronized (ctx) {
```
This existed prior to me.

jhult commented on flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java:
```diff
@@ -361,7 +361,7 @@ public static void initDevModeHandler(Set<Class<?>> classes,
         // Check whether executor is provided by the caller (framework)
         Executor service = lookup.lookup(Executor.class);
```
This existed prior to me.

jhult commented on flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinAppShellInitializer.java:
```diff
@@ -85,12 +83,30 @@ public void process(Set<Class<?>> classes, ServletContext context)
      */
-    @SuppressWarnings("unchecked")
+    @Deprecated
     public static void init(Set<Class<?>> classes, ServletContext context) {
```
Understood.

jhult commented on flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinContextStartupInitializer.java:
```diff
@@ -0,0 +1,71 @@
+     */
+    @Override
+    @Deprecated
+    default void process(Set<Class<?>> classSet, ServletContext ctx) throws ServletException {
```
Understood.

jhult commented:
@pleku and @denis-anisimov,

I think this is finally ready to be merged. All checks/tests are now passing.

Once this is merged, I will submit a PR for the `Fusion` initializer.

jhult commented on flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java:
```diff
@@ -401,11 +431,24 @@ public static void initDevModeHandler(Set<Class<?>> classes,
      *         started, <code>false</code> - otherwise
      */
+    @Deprecated
     public static boolean isDevModeAlreadyStarted(
```
Understood.

jhult commented on flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java:
```diff
@@ -254,9 +260,34 @@ private void setDevModeStarted(ServletContext context) {
      *             if dev mode can't be initialized
      */
+    @Deprecated
     public static void initDevModeHandler(Set<Class<?>> classes,
```
Understood.

jhult reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/vaadin/flow/8469)* status: 9 unresolved discussions, 0 of 1 LGTMs obtained (waiting on @denis-anisimov and @deprecated)

___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/AnnotationValidator.java, line 40 at r2](https://reviewable.io/reviews/vaadin/flow/8469#-MG7D3WT4QGPIW7k-Kn5:-MPe61Nd3UQFwbeRGS4z:b9tes1o) ([raw file](https://github.com/vaadin/flow/blob/c64637e6506472610709bcc91da15cb39e524df5/flow-server/src/main/java/com/vaadin/flow/server/startup/AnnotationValidator.java#L40)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
>  ClassLoaderAwareServletContainerInitializer
> ```

Look like this class may extend `VaadinContextInitializer` directly and all deprecation can be done there instead modifying each class which extends `ClassLoaderAwareServletContainerInitializer`.
</blockquote></details>

I updated `VaadinContextStartupInitializer` to extend `ClassLoaderAwareServletContainerInitializer` and have a default method in `VaadinContextStartupInitializer` which is deprecated.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/ClassLoaderAwareServletContainerInitializer.java, line 90 at r2](https://reviewable.io/reviews/vaadin/flow/8469#-MG7DWyDDtEQxdM_xkJf:-MPe6HIxA9pIfzWGTr-w:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/c64637e6506472610709bcc91da15cb39e524df5/flow-server/src/main/java/com/vaadin/flow/server/startup/ClassLoaderAwareServletContainerInitializer.java#L90)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
> Stream.of(initializer.getDeclaredMethod(processMethodName, Set.class, ServletContext.class))
> ```

This construction is ugly.
There is no need to use `Stream` if the method returns only one method.
In fact `Stream` has been used to avoid an exception which will be thrown if there is no such methods.
Generally : it's better to avoid an exception if it's possible.
So I would suggest to revert this code back and avoid the exception.
Just check all the methods and choose the required one.
</blockquote></details>

Done.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/ConnectEndpointsValidator.java, line 41 at r2](https://reviewable.io/reviews/vaadin/flow/8469#-MG7E8ek5OpOmfOZgxPc:-MPe8ogM4cnjb0Ald2OD:b-4e3rop) ([raw file](https://github.com/vaadin/flow/blob/c64637e6506472610709bcc91da15cb39e524df5/flow-server/src/main/java/com/vaadin/flow/server/startup/ConnectEndpointsValidator.java#L41)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
> ClassLoaderAwareServletContainerInitializer
> ```

Here and everywhere else (as above): `ClassLoaderAwareServletContainerInitializer` should extend `VaadinContextInitializer` directly.
</blockquote></details>

I made `VaadinContextStartupInitializer` extend `VaadinContextStartupInitializer`. Some initializers (such as `ServletVerifier` and `LookupInitializer` need `ServletContext` and not just`VaadinContext`.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java, line 196 at r2](https://reviewable.io/reviews/vaadin/flow/8469#-MG7EWl54ytUHk-qZlII:-MPe6YK80iWIgAUAiJeA:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/c64637e6506472610709bcc91da15cb39e524df5/flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java#L196)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
> DEV_MODE_HANDLER_ALREADY_STARTED_ATTRIBUTE
> ```


There is no any need in this constant in case of using `DevModeInitializer.class`. It can be used inline.
</blockquote></details>

Done.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java, line 600 at r2](https://reviewable.io/reviews/vaadin/flow/8469#-MG7ErX-Aj3-Jsbk_dsW:-MPe6_Li1yIbZACab2xK:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/c64637e6506472610709bcc91da15cb39e524df5/flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java#L600)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
> VaadinInitializerException
> ```

The exception should be checked and declared explicitly in the `throws` of the method .
It should be just consistent with `ServletException`.
</blockquote></details>

Done.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinAppShellInitializer.java, line 72 at r2](https://reviewable.io/reviews/vaadin/flow/8469#-MG7FDf56CElShXeNocs:-MPe6exa1be1xofpuU8g:b-a0xzeb) ([raw file](https://github.com/vaadin/flow/blob/c64637e6506472610709bcc91da15cb39e524df5/flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinAppShellInitializer.java#L72)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
> VaadinAppShellInitializer
> ```

any reason why this class doesn't implement `VaadinContextInitializer` ?
</blockquote></details>

It implements it now.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinInitializerException.java, line 23 at r2](https://reviewable.io/reviews/vaadin/flow/8469#-MG7FbiIC0nn2QuA_8Ra:-MPe79fZ-IWQM9SN7qCr:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/c64637e6506472610709bcc91da15cb39e524df5/flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinInitializerException.java#L23)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
> RuntimeException
> ```

As mentioned above: this exception should be checked.
</blockquote></details>

Done.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinContextInitializer.java, line 43 at r2](https://reviewable.io/reviews/vaadin/flow/8469#-MG7FRIQ0iXdqCmV6rIz:-MPe6sns4CsOQgUfaJor:b-pjfjae) ([raw file](https://github.com/vaadin/flow/blob/c64637e6506472610709bcc91da15cb39e524df5/flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinContextInitializer.java#L43)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
> process
> ```

Minor :
this method may have another name by the way so that it's easier to find it via reflection.
</blockquote></details>

I ended up using `load` as the method name.


<!-- Sent from Reviewable.io -->

denis-anisimov reviewed (changes requested):
> Reviewed 7 of 21 files at r3.
> *[Reviewable](https://reviewable.io/reviews/vaadin/flow/8469)* status: 4 unresolved discussions, 0 of 1 LGTMs obtained (waiting on @denis-anisimov and @deprecated)

___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java, line 198 at r3](https://reviewable.io/reviews/vaadin/flow/8469#-MPrSZhqCk_antirHtE6:-MPrSZhqCk_antirHtE7:b-h6t4a7) ([raw file](https://github.com/vaadin/flow/blob/a75f22f510778329f9e70b568116b3b6e1bf723f/flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java#L198)):*
> ```Java
> 
>     private void setDevModeStarted(VaadinContext context) {
>         context.setAttribute(DevModeInitializer.class, this);
> ```

Is this attribute removed anywhere ?

The problem with `context.setAttribute(DevModeInitializer.class, this)` : context sets the attribute with the dev mode initializer instance and keep it there forever.
`DevModeInitializer`  instance is a short living instance which is created and called in the very beginning of the web app initiailization.
Then it may be GCed. 
Storing the instance onto the context doesn't allow ever to be it GCed.

So the attribute should be either removed at some point or something different should be stored.

Several considerations:
- this PR tries to solve too many things.  It should be just a refactoring but in fact it also changes the logic.
- I would recommend either revert this change back and don't modify the logic or fix it properly (introduce a private nested class and use it in the attribute).
- I don't even see the need of this change at all.
- Making changes in logic and stretching in time this PR introduces dramatic problems in review: I don't remember already anything about this PR , many merge conflict resolutions are required.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinContextInitializer.java, line 43 at r2](https://reviewable.io/reviews/vaadin/flow/8469#-MG7FRIQ0iXdqCmV6rIz:-MPrYezhDBg3Suf5R9kN:b-uosdwu) ([raw file](https://github.com/vaadin/flow/blob/c64637e6506472610709bcc91da15cb39e524df5/flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinContextInitializer.java#L43)):*
<details><summary><i>Previously, jhult (Jonathan Hult) wrote…</i></summary><blockquote>

I ended up using `load` as the method name.
</blockquote></details>

`load`  doesn't say me anything. What does it load ?
Again: in the current design  `VaadinContextStartupInitializer` is just a subinterface of `ServletContainerInitializer`  which doesn't load anything. It's a "listener" method and `load` here  is just a delegate from `ServletContainerInitializer::onStartup`.
So the method should have javadocs similar to `onStartup`  and have some similar name.

In the proper design (when `VaadinContextStartupInitializer` doesn't extend `ClassLoaderAwareServletContainerInitializer`) this method should not know/say anything about `onStartup` but anyway it should have similar javadocs, mention that this is a notification/listener method and have a proper name.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinContextStartupInitializer.java, line 27 at r3](https://reviewable.io/reviews/vaadin/flow/8469#-MPrVvQH7LcH581MZ8v9:-MPrVvQH7LcH581MZ8vA:bwm9uo7) ([raw file](https://github.com/vaadin/flow/blob/a75f22f510778329f9e70b568116b3b6e1bf723f/flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinContextStartupInitializer.java#L27)):*
> ```Java
> 
> /**
>  * Provides a mechanism to load initializers via {@link VaadinContext}.
> ```

It's not clear which initializers.

`ClassLoaderAwareServletContainerInitializer` should be referenced , `ServletContainerInitializer` should be referenced.

It should be mentioned that this is just a facade to delegate calls to `ServletContainerInitializer`.

In fact now I don't even understand the need of this PR at all.
The problem with this approach is: 

`public interface VaadinContextStartupInitializer extends ClassLoaderAwareServletContainerInitializer`.

So you are making not an adapter but this some specialization of `ServletContainerInitializer`.

I don't see why it may be useful.

If you compare similar refactoring for `VaadinContext`  then you may see that `VaadinContext`  doesn't know anything about Servlet. Instead `VaadinServletContext` implements `VaadinContext` .

It's intentionally done in the way to be used in containers which are not Servlet containers (e.g. Portlet).

`VaadinContextStartupInitializer` is still a `ServletContainerInitializer`.
So it works only because of the nature of `ServletContainerInitializer`: a Servlet container calls `ServletContainerInitializer::onStartup` and that triggers the initializer logic : no-one else is calling directly any `load` or `process`  methods.

May be `VaadinContextStartupInitializer` should not extend `ClassLoaderAwareServletContainerInitializer` (at least with the current approach).

Or may be the proper way should be:
- introduce an abstraction `VaadinContextStartupInitializer` which  doesn't extend anything.
- introduce `VaadinServletContextStartupInitializer`  which extends `VaadinContextStartupInitializer` and `ClassLoaderAwareServletContainerInitializer` with the implementation here.

Every existing servlet initiiaizer then should implement `VaadinServletContextStartupInitializer` .


<!-- Sent from Reviewable.io -->

denis-anisimov reviewed (changes requested):
> Reviewed 13 of 21 files at r3.
> *[Reviewable](https://reviewable.io/reviews/vaadin/flow/8469)* status: 13 unresolved discussions, 0 of 1 LGTMs obtained (waiting on @denis-anisimov and @deprecated)

___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/ApplicationRouteRegistry.java, line 105 at r3](https://reviewable.io/reviews/vaadin/flow/8469#-MPra6YO5KuDKKpq9eqQ:-MPra6YO5KuDKKpq9eqR:bjwgwvp) ([raw file](https://github.com/vaadin/flow/blob/a75f22f510778329f9e70b568116b3b6e1bf723f/flow-server/src/main/java/com/vaadin/flow/server/startup/ApplicationRouteRegistry.java#L105)):*
> ```Java
>      *
>      * @param context
>      *            the {@link VaadinContext} for which to get a <tt>RouteRegistry</tt>, not
> ```

please revert this.
`@param` doesn't support `link`  .
the result will be broken javadocs for param

And please avoid extra changes for the PR.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/ApplicationRouteRegistry.java, line 107 at r3](https://reviewable.io/reviews/vaadin/flow/8469#-MPra7wW3beaQLd_OIR8:-MPra7wW3beaQLd_OIR9:b-v4kghq) ([raw file](https://github.com/vaadin/flow/blob/a75f22f510778329f9e70b568116b3b6e1bf723f/flow-server/src/main/java/com/vaadin/flow/server/startup/ApplicationRouteRegistry.java#L107)):*
> ```Java
>      *            the {@link VaadinContext} for which to get a <tt>RouteRegistry</tt>, not
>      *            <code>null</code>
>      * @return an  for the given {@link VaadinContext}, not
> ```

please revert this.
`@return` doesn't support `link`  .
the result will be broken javadocs for param

And please avoid extra changes for the PR.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/DeferredServletContextInitializers.java, line 55 at r3](https://reviewable.io/reviews/vaadin/flow/8469#-MPra5K4ChTQN9NUQtOl:-MPra5K4ChTQN9NUQtOm:bjwgwvp) ([raw file](https://github.com/vaadin/flow/blob/a75f22f510778329f9e70b568116b3b6e1bf723f/flow-server/src/main/java/com/vaadin/flow/server/startup/DeferredServletContextInitializers.java#L55)):*
> ```Java
>          * 
>          * @param context
>          *            a {@link ServletContext} for the initializer
> ```

please revert this.
`@param` doesn't support `link`  .
the result will be broken javadocs for param

And please avoid extra changes for the PR.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/DeferredServletContextInitializers.java, line 78 at r3](https://reviewable.io/reviews/vaadin/flow/8469#-MPr_rWwE71QxmFtycq6:-MPr_rWwE71QxmFtycq7:bjwgwvp) ([raw file](https://github.com/vaadin/flow/blob/a75f22f510778329f9e70b568116b3b6e1bf723f/flow-server/src/main/java/com/vaadin/flow/server/startup/DeferredServletContextInitializers.java#L78)):*
> ```Java
>      * 
>      * @param context
>      *            a {@link ServletContext} for initializers
> ```

please revert this.
`@param` doesn't support `link`  .
the result will be broken javadocs for param

And please avoid extra changes for the PR.
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/LookupInitializer.java, line 276 at r3](https://reviewable.io/reviews/vaadin/flow/8469#-MPraIAT3qjg7-2zeUBJ:-MPraIAT3qjg7-2zeUBK:b-dfwip) ([raw file](https://github.com/vaadin/flow/blob/a75f22f510778329f9e70b568116b3b6e1bf723f/flow-server/src/main/java/com/vaadin/flow/server/startup/LookupInitializer.java#L276)):*
> ```Java
>      *            the Set of application classes that extend, implement, or have
>      *            been annotated with the class types specified by the
>      *            {@link javax.servlet.annotation.HandlesTypes}
> ```

Don't use link inside `@param`
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/LookupInitializer.java, line 281 at r3](https://reviewable.io/reviews/vaadin/flow/8469#-MPraNdK0XNc6TKkhl1d:-MPraNdK0XNc6TKkhl1e:b-dfwip) ([raw file](https://github.com/vaadin/flow/blob/a75f22f510778329f9e70b568116b3b6e1bf723f/flow-server/src/main/java/com/vaadin/flow/server/startup/LookupInitializer.java#L281)):*
> ```Java
>      *
>      * @param context
>      *            the {@link VaadinContext} to use with initializer
> ```

Don't use link inside `@param`
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/ServletDeployer.java, line 258 at r3](https://reviewable.io/reviews/vaadin/flow/8469#-MPraS9b0OxUb2S7mOuI:-MPraS9b0OxUb2S7mOuJ:byn7w8l) ([raw file](https://github.com/vaadin/flow/blob/a75f22f510778329f9e70b568116b3b6e1bf723f/flow-server/src/main/java/com/vaadin/flow/server/startup/ServletDeployer.java#L258)):*
> ```Java
>      *
>      * @param servletContext
>      *            the deployed {@link ServletContext}
> ```

please revert
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinContextStartupInitializer.java, line 41 at r3](https://reviewable.io/reviews/vaadin/flow/8469#-MPraZI44i_hdVxHFdGJ:-MPraZI44i_hdVxHFdGK:b-ub4n8t) ([raw file](https://github.com/vaadin/flow/blob/a75f22f510778329f9e70b568116b3b6e1bf723f/flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinContextStartupInitializer.java#L41)):*
> ```Java
>      *            the Set of application classes that extend, implement, or have
>      *            been annotated with the class types specified by the
>      *            {@link javax.servlet.annotation.HandlesTypes}
> ```

link in param
___
*[flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinContextStartupInitializer.java, line 46 at r3](https://reviewable.io/reviews/vaadin/flow/8469#-MPrab4J5HaGv8c9231-:-MPrab4J5HaGv8c92310:b-ub4n8t) ([raw file](https://github.com/vaadin/flow/blob/a75f22f510778329f9e70b568116b3b6e1bf723f/flow-server/src/main/java/com/vaadin/flow/server/startup/VaadinContextStartupInitializer.java#L46)):*
> ```Java
>      *
>      * @param context
>      *            the {@link VaadinContext} to use with initializer
> ```

link in param


<!-- Sent from Reviewable.io -->

denis-anisimov commented:
This PR has quite many chances never be merged: it contains too many changes which are unrelated to refactoring.
Also every new update introduces more unrelated changes, it's too hard to track such changes and it's extremely stretched out in time.

jhult commented:
I will submit a clean pull request (taking your notes into consideration) and close this one.

FYI, using `@link` in `@param` appears to be fair game as it is done in the following linked examples:

- Core Java
	- [Exception](https://github.com/AdoptOpenJDK/openjdk-jdk16/blob/master/src/java.base/share/classes/java/lang/Exception.java#L64)
	- [CompactNumberFormat](https://github.com/AdoptOpenJDK/openjdk-jdk16/blob/master/src/java.base/share/classes/java/text/CompactNumberFormat.java#L481)
- Guava
	- [MoreObjects]( https://github.com/google/guava/blob/master/guava/src/com/google/common/base/MoreObjects.java#L121)
	- [Verify](https://github.com/google/guava/blob/master/guava/src/com/google/common/base/Verify.java#L115)
- Vaadin Flow itself
	- [RouteAliasData](https://github.com/vaadin/flow/blob/master/flow-server/src/main/java/com/vaadin/flow/router/RouteAliasData.java#L45)
	- [AbstractTheme](https://github.com/vaadin/flow/blob/master/flow-server/src/main/java/com/vaadin/flow/theme/AbstractTheme.java#L74)
	- [BootstrapException](https://github.com/vaadin/flow/blob/master/flow-server/src/main/java/com/vaadin/flow/server/BootstrapException.java#L52)

In addition, Oracle themselves states the following:

1. > Notice the inline tag `{@link URL}`, which converts to an HTML hyperlink pointing to the documentation for the URL class. This inline tag can be used anywhere that a comment can be written, such as in the text following block tags. 
[Source](https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html)
2. > Inserts an inline link with a visible text label that points to the documentation for the specified package, class, or member name of a referenced class. This tag is valid in all documentation comments: overview, package, class, interface, constructor, method and field, including the text portion of any tag, such as the `@return`, `@param` and `@deprecated` tags.
[Source](https://docs.oracle.com/javase/8/docs/technotes/tools/windows/javadoc.html#CHDDIECH)

denis-anisimov commented:
> I will submit a clean pull request (taking your notes into consideration) and close this one.
> 
> FYI, using `@link` in `@param` appears to be fair game as it is done in the following linked examples:
> 
> * Core Java
>   
>   * [Exception](https://github.com/AdoptOpenJDK/openjdk-jdk16/blob/master/src/java.base/share/classes/java/lang/Exception.java#L64)
>   * [CompactNumberFormat](https://github.com/AdoptOpenJDK/openjdk-jdk16/blob/master/src/java.base/share/classes/java/text/CompactNumberFormat.java#L481)
> * Guava
>   
>   * [MoreObjects](https://github.com/google/guava/blob/master/guava/src/com/google/common/base/MoreObjects.java#L121)
>   * [Verify](https://github.com/google/guava/blob/master/guava/src/com/google/common/base/Verify.java#L115)
> * Vaadin Flow itself
>   
>   * [RouteAliasData](https://github.com/vaadin/flow/blob/master/flow-server/src/main/java/com/vaadin/flow/router/RouteAliasData.java#L45)
>   * [AbstractTheme](https://github.com/vaadin/flow/blob/master/flow-server/src/main/java/com/vaadin/flow/theme/AbstractTheme.java#L74)
>   * [BootstrapException](https://github.com/vaadin/flow/blob/master/flow-server/src/main/java/com/vaadin/flow/server/BootstrapException.java#L52)
> 
> In addition, Oracle themselves states the following:
> 
> 1. > Notice the inline tag `{@link URL}`, which converts to an HTML hyperlink pointing to the documentation for the URL class. This inline tag can be used anywhere that a comment can be written, such as in the text following block tags.
>    > [Source](https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html)
> 2. > Inserts an inline link with a visible text label that points to the documentation for the specified package, class, or member name of a referenced class. This tag is valid in all documentation comments: overview, package, class, interface, constructor, method and field, including the text portion of any tag, such as the `@return`, `@param` and `@deprecated` tags.
>    > [Source](https://docs.oracle.com/javase/8/docs/technotes/tools/windows/javadoc.html#CHDDIECH)

Looks like you are right.
My bad.
Either there was an issue previously with this and it's fixed now or I have problems with my memory.
I've been avoiding at least using links in the `@param` all the time.
But it's good to know that it's actually works.
Thanks.
